File: coprocessor/frc/python/video_to_disk/video_to_disk.py
#!/usr/bin/env python3

# Since ROS adds python.27 packages onto the system path, remove this when running this python3 code.
import sys
sys.path.remove('/opt/ros/kinetic/lib/python2.7/dist-packages')

import cscore as cs
import cv2 as cv
import numpy as np

width=320
height=240
frames_per_sec=15

camera = cs.UsbCamera("usbcam", 0)
camera.setVideoMode(cs.VideoMode.PixelFormat.kMJPEG, width, height, frames_per_sec)

mjpegServer = cs.MjpegServer("httpserver", 8081)
mjpegServer.setSource(camera)

cvsink = cs.CvSink("cvsink")
cvsink.setSource(camera)

cvsource = cs.CvSource("cvsource", cs.VideoMode.PixelFormat.kMJPEG, width, height, frames_per_sec)

cvMjpegServer = cs.MjpegServer("cvhttpserver", 8082)
cvMjpegServer.setSource(cvsource)

frameSize = (width, height)
videoWriter = cv.VideoWriter("/media/pi/data/output.avi",
				cv.VideoWriter.fourcc('F', 'M', 'P', '4'), 15.0, frameSize, True)

img = np.zeros(shape=(height, width, 3), dtype=np.uint8)    

count = 0
while count < 100:
	time = cvsink.grabFrame(img)
	print("Timestamp: %d" % time[0])
	videoWriter.write(img)
	count = count + 1

File: coprocessor/frc/cpp/video_to_disk/video_to_disk.cpp
#include <cscore.h>
#include <opencv2/core/core.hpp>
#include <opencv2/videoio/videoio.hpp>
#include <iostream>

using namespace cs;
using namespace cv;

int main() {
	UsbCamera camera("usbcam", 0);
	camera.SetVideoMode(VideoMode::PixelFormat::kMJPEG, 320, 240, 15);
	MjpegServer mjpegServer("httpserver", 8081);
	mjpegServer.SetSource(camera);
	CvSink cvsink("cvsink");
	cvsink.SetSource(camera);
	CvSource cvsource("cvsource",
			VideoMode::PixelFormat::kMJPEG, 320, 240, 15);
	MjpegServer cvMjpegServer("cvhttpserver", 8082);
	cvMjpegServer.SetSource(cvsource);

	Size frameSize(320, 240);
	VideoWriter videoWriter("/data/output.avi",
			VideoWriter::fourcc('F', 'M', 'P', '4'), 15.0, frameSize, true);

	Mat frame;

	int count = 0;
	while (count < 100) {
		uint64_t time = cvsink.GrabFrame(frame);
		printf("Frame Time:  %lld\n", time);
		videoWriter.write(frame);
		count++;
	}
}
